File: uis/client/web2/src/services/reward_service.ts
import { AppConfig } from '../config/app_config';
import { apiService } from './api_service';

export interface Reward {
  id: string;
  userId: string;
  title: string;
  description: string;
  points: number;
  category: 'cashback' | 'bonus' | 'milestone' | 'referral' | 'streak';
  date: Date;
  transactionId?: string;
}

export interface RewardSummary {
  totalPoints: number;
  earnedThisMonth: number;
  redeemedThisMonth: number;
  tier: string;
  nextTierPoints: number;
  lifetimeEarned: number;
  lifetimeRedeemed: number;
}

export interface RedeemOption {
  id: string;
  title: string;
  description: string;
  pointsRequired: number;
  category: string;
  imageUrl?: string;
  isAvailable: boolean;
}

export interface RedemptionHistory {
  id: string;
  optionId: string;
  title: string;
  pointsSpent: number;
  date: Date;
  status: 'pending' | 'completed' | 'failed';
  trackingNumber?: string;
}

export class RewardService {
  // Get user rewards summary
  async getRewardSummary(): Promise<RewardSummary> {
    try {
      const response = await apiService.get(`${AppConfig.rewardEndpoint}/summary`);
      
      const data = response.data as { success?: boolean; data?: any };
      if (data.success || response.status === 200) {
        const summary = data.data || {};
        return {
          totalPoints: summary.total_points || 0,
          earnedThisMonth: summary.earned_this_month || 0,
          redeemedThisMonth: summary.redeemed_this_month || 0,
          tier: summary.tier || 'Bronze',
          nextTierPoints: summary.next_tier_points || 0,
          lifetimeEarned: summary.lifetime_earned || 0,
          lifetimeRedeemed: summary.lifetime_redeemed || 0,
        };
      }
      return {
        totalPoints: 0,
        earnedThisMonth: 0,
        redeemedThisMonth: 0,
        tier: 'Bronze',
        nextTierPoints: 1000,
        lifetimeEarned: 0,
        lifetimeRedeemed: 0,
      };
    } catch (error) {
      console.error('Failed to fetch reward summary:', error);
      return {
        totalPoints: 0,
        earnedThisMonth: 0,
        redeemedThisMonth: 0,
        tier: 'Bronze',
        nextTierPoints: 1000,
        lifetimeEarned: 0,
        lifetimeRedeemed: 0,
      };
    }
  }

  // Get earned rewards history
  async getEarnedRewards(): Promise<Reward[]> {
    try {
      const response = await apiService.get(`${AppConfig.rewardEndpoint}/earned`);
      
      const data = response.data as { success?: boolean; data?: Record<string, unknown>[] };
      if (data.success || response.status === 200) {
        const rewardsData = data.data as Record<string, unknown>[];
        return rewardsData.map((reward) => ({
          id: reward.id as string,
          userId: reward.user_id as string,
          title: reward.title as string,
          description: reward.description as string,
          points: reward.points as number,
          category: reward.category as Reward['category'],
          date: new Date(reward.date as string),
          transactionId: reward.transaction_id as string | undefined,
        }));
      }
      return [];
    } catch (error) {
      console.error('Failed to fetch earned rewards:', error);
      return [];
    }
  }

  // Get available redemption options
  async getRedeemOptions(): Promise<RedeemOption[]> {
    try {
      const response = await apiService.get(`${AppConfig.rewardEndpoint}/redeem-options`);
      
      const data = response.data as { success?: boolean; data?: Record<string, unknown>[] };
      if (data.success || response.status === 200) {
        const optionsData = data.data as Record<string, unknown>[];
        return optionsData.map((option) => ({
          id: option.id as string,
          title: option.title as string,
          description: option.description as string,
          pointsRequired: option.points_required as number,
          category: option.category as string,
          imageUrl: option.image_url as string | undefined,
          isAvailable: option.is_available as boolean,
        }));
      }
      return [];
    } catch (error) {
      console.error('Failed to fetch redeem options:', error);
      return [];
    }
  }

  // Redeem reward points
  async redeemPoints(optionId: string, deliveryInfo?: Record<string, unknown>): Promise<{ success: boolean; message: string; data?: RedemptionHistory }> {
    try {
      const response = await apiService.post(`${AppConfig.rewardEndpoint}/redeem`, {
        option_id: optionId,
        delivery_info: deliveryInfo,
      });

      const data = response.data as { success?: boolean; message?: string; data?: any };
      if (data.success || response.status === 200) {
        return {
          success: true,
          message: data.message || 'Reward redeemed successfully',
          data: data.data ? {
            id: data.data.id,
            optionId: data.data.option_id,
            title: data.data.title,
            pointsSpent: data.data.points_spent,
            date: new Date(data.data.date),
            status: data.data.status,
            trackingNumber: data.data.tracking_number,
          } : undefined,
        };
      }
      return {
        success: false,
        message: data.message || 'Failed to redeem reward',
      };
    } catch (error: unknown) {
      return {
        success: false,
        message: error instanceof Error ? error.message : 'Error redeeming reward',
      };
    }
  }

  // Get redemption history
  async getRedemptionHistory(): Promise<RedemptionHistory[]> {
    try {
      const response = await apiService.get(`${AppConfig.rewardEndpoint}/redemption-history`);
      
      const data = response.data as { success?: boolean; data?: Record<string, unknown>[] };
      if (data.success || response.status === 200) {
        const historyData = data.data as Record<string, unknown>[];
        return historyData.map((item) => ({
          id: item.id as string,
          optionId: item.option_id as string,
          title: item.title as string,
          pointsSpent: item.points_spent as number,
          date: new Date(item.date as string),
          status: item.status as RedemptionHistory['status'],
          trackingNumber: item.tracking_number as string | undefined,
        }));
      }
      return [];
    } catch (error) {
      console.error('Failed to fetch redemption history:', error);
      return [];
    }
  }

  // Get rewards by category
  async getRewardsByCategory(category: string): Promise<Reward[]> {
    try {
      const response = await apiService.get(`${AppConfig.rewardEndpoint}/earned?category=${category}`);
      
      const data = response.data as { success?: boolean; data?: Record<string, unknown>[] };
      if (data.success || response.status === 200) {
        const rewardsData = data.data as Record<string, unknown>[];
        return rewardsData.map((reward) => ({
          id: reward.id as string,
          userId: reward.user_id as string,
          title: reward.title as string,
          description: reward.description as string,
          points: reward.points as number,
          category: reward.category as Reward['category'],
          date: new Date(reward.date as string),
          transactionId: reward.transaction_id as string | undefined,
        }));
      }
      return [];
    } catch (error) {
      console.error('Failed to fetch rewards by category:', error);
      return [];
    }
  }
}

export const rewardService = new RewardService();

File: uis/client/mobile_app/lib/services/reward_service.dart
import '../config/app_config.dart';
import 'api_service.dart';

// Type definitions
class Reward {
  final String id;
  final String userId;
  final String title;
  final String description;
  final int points;
  final String category; // 'cashback' | 'bonus' | 'milestone' | 'referral' | 'streak'
  final DateTime date;
  final String? transactionId;

  Reward({
    required this.id,
    required this.userId,
    required this.title,
    required this.description,
    required this.points,
    required this.category,
    required this.date,
    this.transactionId,
  });

  factory Reward.fromJson(Map<String, dynamic> json) {
    return Reward(
      id: json['id'] as String,
      userId: json['user_id'] as String,
      title: json['title'] as String,
      description: json['description'] as String,
      points: json['points'] as int,
      category: json['category'] as String,
      date: DateTime.parse(json['date'] as String),
      transactionId: json['transaction_id'] as String?,
    );
  }
}

class RewardSummary {
  final int totalPoints;
  final int earnedThisMonth;
  final int redeemedThisMonth;
  final String tier;
  final int nextTierPoints;
  final int lifetimeEarned;
  final int lifetimeRedeemed;

  RewardSummary({
    required this.totalPoints,
    required this.earnedThisMonth,
    required this.redeemedThisMonth,
    required this.tier,
    required this.nextTierPoints,
    required this.lifetimeEarned,
    required this.lifetimeRedeemed,
  });
}

class RedeemOption {
  final String id;
  final String title;
  final String description;
  final int pointsRequired;
  final String category;
  final String? imageUrl;
  final bool isAvailable;

  RedeemOption({
    required this.id,
    required this.title,
    required this.description,
    required this.pointsRequired,
    required this.category,
    this.imageUrl,
    required this.isAvailable,
  });

  factory RedeemOption.fromJson(Map<String, dynamic> json) {
    return RedeemOption(
      id: json['id'] as String,
      title: json['title'] as String,
      description: json['description'] as String,
      pointsRequired: json['points_required'] as int,
      category: json['category'] as String,
      imageUrl: json['image_url'] as String?,
      isAvailable: json['is_available'] as bool,
    );
  }
}

class RedemptionHistory {
  final String id;
  final String optionId;
  final String title;
  final int pointsSpent;
  final DateTime date;
  final String status; // 'pending' | 'completed' | 'failed'
  final String? trackingNumber;

  RedemptionHistory({
    required this.id,
    required this.optionId,
    required this.title,
    required this.pointsSpent,
    required this.date,
    required this.status,
    this.trackingNumber,
  });

  factory RedemptionHistory.fromJson(Map<String, dynamic> json) {
    return RedemptionHistory(
      id: json['id'] as String,
      optionId: json['option_id'] as String,
      title: json['title'] as String,
      pointsSpent: json['points_spent'] as int,
      date: DateTime.parse(json['date'] as String),
      status: json['status'] as String,
      trackingNumber: json['tracking_number'] as String?,
    );
  }
}

class RewardService {
  final ApiService _apiService = ApiService();

  // Get user rewards summary
  Future<RewardSummary> getRewardSummary() async {
    try {
      final response = await _apiService.get('${AppConfig.rewardEndpoint}/summary');

      if (response.data['success'] == true || response.statusCode == 200) {
        final data = response.data['data'] as Map<String, dynamic>;
        return RewardSummary(
          totalPoints: data['total_points'] ?? 0,
          earnedThisMonth: data['earned_this_month'] ?? 0,
          redeemedThisMonth: data['redeemed_this_month'] ?? 0,
          tier: data['tier'] ?? 'Bronze',
          nextTierPoints: data['next_tier_points'] ?? 0,
          lifetimeEarned: data['lifetime_earned'] ?? 0,
          lifetimeRedeemed: data['lifetime_redeemed'] ?? 0,
        );
      }

      // Return default summary if API fails
      return RewardSummary(
        totalPoints: 0,
        earnedThisMonth: 0,
        redeemedThisMonth: 0,
        tier: 'Bronze',
        nextTierPoints: 1000,
        lifetimeEarned: 0,
        lifetimeRedeemed: 0,
      );
    } catch (error) {
      print('Failed to fetch reward summary: $error');
      return RewardSummary(
        totalPoints: 0,
        earnedThisMonth: 0,
        redeemedThisMonth: 0,
        tier: 'Bronze',
        nextTierPoints: 1000,
        lifetimeEarned: 0,
        lifetimeRedeemed: 0,
      );
    }
  }

  // Get earned rewards history
  Future<List<Reward>> getEarnedRewards() async {
    try {
      final response = await _apiService.get('${AppConfig.rewardEndpoint}/earned');

      if (response.data['success'] == true || response.statusCode == 200) {
        final rewardsData = response.data['data'] as List<dynamic>;
        return rewardsData.map((reward) => Reward.fromJson(reward as Map<String, dynamic>)).toList();
      }

      return [];
    } catch (error) {
      print('Failed to fetch earned rewards: $error');
      return [];
    }
  }

  // Get available redemption options
  Future<List<RedeemOption>> getRedeemOptions() async {
    try {
      final response = await _apiService.get('${AppConfig.rewardEndpoint}/redeem-options');

      if (response.data['success'] == true || response.statusCode == 200) {
        final optionsData = response.data['data'] as List<dynamic>;
        return optionsData.map((option) => RedeemOption.fromJson(option as Map<String, dynamic>)).toList();
      }

      return [];
    } catch (error) {
      print('Failed to fetch redeem options: $error');
      return [];
    }
  }

  // Redeem reward points
  Future<Map<String, dynamic>> redeemPoints(String optionId, {Map<String, dynamic>? deliveryInfo}) async {
    try {
      final response = await _apiService.post(
        '${AppConfig.rewardEndpoint}/redeem',
        data: {
          'option_id': optionId,
          if (deliveryInfo != null) 'delivery_info': deliveryInfo,
        },
      );

      if (response.data['success'] == true || response.statusCode == 200) {
        return {
          'success': true,
          'message': response.data['message'] ?? 'Reward redeemed successfully',
          'data': response.data['data'] != null
              ? RedemptionHistory.fromJson(response.data['data'] as Map<String, dynamic>)
              : null,
        };
      }

      return {
        'success': false,
        'message': response.data['message'] ?? 'Failed to redeem reward',
      };
    } catch (error) {
      return {
        'success': false,
        'message': error.toString(),
      };
    }
  }

  // Get redemption history
  Future<List<RedemptionHistory>> getRedemptionHistory() async {
    try {
      final response = await _apiService.get('${AppConfig.rewardEndpoint}/redemption-history');

      if (response.data['success'] == true || response.statusCode == 200) {
        final historyData = response.data['data'] as List<dynamic>;
        return historyData.map((item) => RedemptionHistory.fromJson(item as Map<String, dynamic>)).toList();
      }

      return [];
    } catch (error) {
      print('Failed to fetch redemption history: $error');
      return [];
    }
  }

  // Get rewards by category
  Future<List<Reward>> getRewardsByCategory(String category) async {
    try {
      final response = await _apiService.get('${AppConfig.rewardEndpoint}/earned?category=$category');

      if (response.data['success'] == true || response.statusCode == 200) {
        final rewardsData = response.data['data'] as List<dynamic>;
        return rewardsData.map((reward) => Reward.fromJson(reward as Map<String, dynamic>)).toList();
      }

      return [];
    } catch (error) {
      print('Failed to fetch rewards by category: $error');
      return [];
    }
  }
}

